src/codegen/statements.rs

```diff
@@ -946,31 +950,54 @@ fn destructure(
                 );
             }
             DestructureField::Expression(left) => {
-                // the resolver did not cast the expression
-                let loc = left.loc();
-
-                let expr = cast(
-                    &loc,
-                    right,
-                    left.ty().deref_any(),
-                    true,
-                    ns,
-                    &mut Vec::new(),
-                )
-                .expect("sema should have checked cast");
-                // casts to StorageLoad generate LoadStorage instructions
-                let expr = expression(&expr, cfg, contract_no, Some(func), ns, vartab, opt);
+                let expr = destructure_load(
+                    &left.loc(),
+                    &right.cast(left.ty().deref_any(), ns),
+                    cfg,
+                    vartab,
+                );
 
                 if should_remove_assignment(ns, left, func, opt) {
                     continue;
                 }
 
-                assign_single(left, &expr, cfg, contract_no, Some(func), ns, vartab, opt);
+                assign_single(left, expr, cfg, contract_no, Some(func), ns, vartab, opt);
             }
         }
     }
 }
 
+/// If the destructure statement returns a storage reference,
+/// we must add a load instruction in the CFG.
```

Comment:
This needs an explanation. For normal assignments, sema will have cast the value to the correct type. However we do not for destructure statements.
```
contract c {
   int[2] x;
    function x() public returns (int[2] storage, int8) { return (x, 1); }
   function y()  public {
       (int[2] a, int16 b) =  x();
   }
}
```
The cast of `int8` to `int16` and from `int[2] storage` to `int[2]` is not stored in the ast. sema only checks that the cast can be done.

The cast in codegen does not insert `Instr::LoadStorage`.